hbase-common/src/main/java/org/apache/hadoop/hbase/util/ReflectedFunctionCache.java

```diff
@@ -0,0 +1,121 @@
+ * all subclasses in the same package of the passed baseClass. For each found class, creates a
+ * lambda using
+ * {@link ReflectionUtils#getOneArgStaticMethodAsFunction(Class, String, Class, Class)}. These are
+ * added to a hashmap for fast lookup by name later.
+ * @param <I> the input argument type for the resolved functions
+ * @param <R> the return type for the resolved functions
+ */
+@InterfaceAudience.Private
+final public class ReflectedFunctionCache<I, R> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ReflectedFunctionCache.class);
+
+  private final Map<String, Function<I, ? extends R>> lambdasByClass;
+
+  private ReflectedFunctionCache(Map<String, Function<I, ? extends R>> lambdasByClass) {
+    this.lambdasByClass = lambdasByClass;
+  }
+
+  /**
+   * Create a cache of reflected functions using the provided classloader and baseClass. Will find
+   * all subclasses of the provided baseClass (in the same package), and then foreach look for a
+   * static one-arg method with the methodName and argClass. The expectation is that the method
+   * returns a value whose class extends the baseClass. This was primarily designed for use by our
+   * Filter and Comparator parseFrom methods.
+   */
+  public static <I, R> ReflectedFunctionCache<I, R> create(ClassLoader classLoader,
+    Class<R> baseClass, Class<I> argClass, String methodName) {
+    Map<String, Function<I, ? extends R>> lambdasByClass = new HashMap<>();
+    Set<? extends Class<? extends R>> classes = getSubclassesInPackage(classLoader, baseClass);
+    for (Class<? extends R> clazz : classes) {
+      Function<I, ? extends R> func = createFunction(clazz, methodName, argClass, clazz);
+      if (func != null) {
+        lambdasByClass.put(clazz.getName(), func);
+      }
+    }
+    return new ReflectedFunctionCache<>(lambdasByClass);
+  }
+
+  /**
+   * Get and execute the Function for the given className, passing the argument to the function and
+   * returning the result.
+   * @param className the full name of the class to lookup
+   * @param argument  the argument to pass to the function, if found.
+   * @return null if a function is not found for classname, otherwise the result of the function.
+   */
+  public R getAndCallByName(String className, I argument) {
+    Function<I, ? extends R> lambda = lambdasByClass.get(className);
+
+    // todo: if we ever make java9+ our lowest supported jdk version, we can
+    // handle generating these for newly loaded classes from our DynamicClassLoader using
+    // MethodHandles.privateLookupIn(). For now this is not possible, because we can't easily
+    // create a privileged lookup in a non-default ClassLoader.
+    if (lambda == null) {
+      return null;
+    }
+
+    return lambda.apply(argument);
+  }
+
+  private static <R> Set<Class<? extends R>> getSubclassesInPackage(ClassLoader classLoader,
+    Class<R> baseClass) {
+    try {
+      return ClassPath.from(classLoader).getAllClasses().stream()
```

Comment:
Will this cause we load all classes even if it is not used by now? I used to use guava's ClassPath in a project but it performed differently when executing in IDE and in command line, finally I chose to use ClassPathScanningCandidateComponentProvider in spring for scanning classes...

Reply:
It will cause us to load all of the matching ones, i.e. the ones i call `load` on below after filtering to the correct package.  So it will load all of the filters in `org.apache.hadoop.hbase.filter` on startup.

I thought this was preferable because the number of classes is not large. Since I do it on startup, I don't need to worry about synchronization. I could only populate the cache as Filters are accessed, but then I need to handle synchronization. Would you prefer that?

Reply:
Reading the code for `ClassPath#getAllClasses`, it seems that it will build an exhaustive set of all classes on the classpath that are loadable. It doesn't actually load the classes. My understanding is that `Stream` operations are lazy, so the only classes loaded should be those that materialize in the final `collect`.